### PR TITLE
Added changes for boots to have centralized logging in tinkerbell

### PR DESCRIPTION
**Pull request by Cbkhare**
Signed-off-by: cbkhare <Chitrabasukhare89@gmail.com>

## Description

Changes to add centralized logging for the Tinkerbell. 

## Why is this needed

- We need a log aggregator for Tinkerbell which can collect log from tink services and tink worker containers. 
- Tink PR: https://github.com/tinkerbell/tink/pull/267
- osie PR: https://github.com/tinkerbell/osie/pull/104
- RFD: https://github.com/tinkerbell/proposals/pull/7

Fixes: #

## How Has This Been Tested?
- [x] Tested with basic hello-world example.
- [x] Verified the testing of logrotation.
- [x] Added below test results. 

**log_aggregation**
```
vagrant@provisioner:/vagrant/deploy$ ls -lrt   /var/log/dockerlfs/
total 20
drwxr-xr-x 2 syslog syslog 4096 Sep  8 12:59 provisioner
-rw-r----- 1 syslog adm    9521 Sep  8 13:04 daemon.log
drwxr-xr-x 2 syslog syslog 4096 Sep  8 13:17 192.168.1.5
vagrant@provisioner:/vagrant/deploy$ ls -lrt   /var/log/dockerlfs/192.168.1.5/
total 12
-rw-r----- 1 syslog adm 2985 Sep  8 13:17 hello_world_0e65404d-5abe-4746-b6a9-bba9be867884.log
-rw-r----- 1 syslog adm 4789 Sep  8 13:17 tink-worker.log
```

**logrotation**:
```
vagrant@provisioner:/vagrant/deploy$ ls -lrt   /var/log/dockerlfs/provisioner/
total 36
-rw-r----- 1 syslog adm  513 Sep  8 13:04 deploy_db_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm 4099 Sep  8 13:17 deploy_boots_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm  577 Sep  8 13:17 deploy_nginx_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm  400 Sep  8 13:17 deploy_hegel_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm 1450 Sep  8 13:17 deploy_tink-server_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm 6155 Sep  8 13:37 deploy_registry_1.log-20200908131599572256.gz
-rw-r----- 1 syslog adm    0 Sep  8 13:37 deploy_boots_1.log
-rw-r----- 1 syslog adm    0 Sep  8 13:37 deploy_db_1.log
-rw-r----- 1 syslog adm    0 Sep  8 13:37 deploy_hegel_1.log
-rw-r----- 1 syslog adm    0 Sep  8 13:37 deploy_nginx_1.log
-rw-r----- 1 syslog adm    0 Sep  8 13:37 deploy_tink-server_1.log
-rw-r----- 1 syslog adm  147 Sep  8 13:37 deploy_registry_1.log
vagrant@provisioner:/vagrant/deploy$ ls -lrt   /var/log/dockerlfs/192.168.1.5/
total 8
-rw-r----- 1 syslog adm 571 Sep  8 13:17 hello_world_0e65404d-5abe-4746-b6a9-bba9be867884.log-20200908131599572256.gz
-rw-r----- 1 syslog adm 918 Sep  8 13:17 tink-worker.log-20200908131599572256.gz
-rw-r----- 1 syslog adm   0 Sep  8 13:37 hello_world_0e65404d-5abe-4746-b6a9-bba9be867884.log
-rw-r----- 1 syslog adm   0 Sep  8 13:37 tink-worker.log
vagrant@provisioner:/vagrant/deploy$ ls -lrt   /var/log/dockerlfs/
total 12
-rw-r----- 1 syslog adm    1642 Sep  8 13:04 daemon.log-20200908131599572256.gz
-rw-r----- 1 syslog adm       0 Sep  8 13:37 daemon.log
drwxr-xr-x 2 syslog syslog 4096 Sep  8 13:37 192.168.1.5
drwxr-xr-x 2 syslog syslog 4096 Sep  8 13:37 provisioner

``` 
## How are existing users impacted? What migration steps/scripts do we need?

- Log location has been updated. Documentation will need an update.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)

### CODE REVIEW

File: installers/osie/mirror.go [hunk 1]
```diff
@@ -14,11 +14,13 @@ const (
...
-	dockerRegistry                     string
-	grpcAuthority, grpcCertURL         string
-	registryUsername, registryPassword string
+	osieURL                                 = mustBuildOSIEURL().String()
+	mirrorBaseURL                           = conf.MirrorBaseUrl
+	dockerRegistry                          string
+	grpcAuthority, grpcCertURL              string
+	registryUsername, registryPassword      string
+	log_driver, log_tag, log_server_address string
+	log_server_address_type                 string
```
Comment (mmlb) [inline installers/osie/mirror.go:23]: not sure why these are multiple vars in one line, especially log_server_address and log_server_address_type in different lines.

File: installers/osie/main.go [hunk 2]
```diff
@@ -81,6 +81,10 @@ func kernelParams(action, state string, j job.Job, s *ipxe.Script) {
 		s.Args("grpc_cert_url=" + grpcCertURL)
 		s.Args("registry_username=" + registryUsername)
 		s.Args("registry_password=" + registryPassword)
+		s.Args("log_driver=" + log_driver)
+		s.Args("log_tag=" + log_tag)
+		s.Args("log_server_address=" + log_server_address)
+		s.Args("log_server_address_type=" + log_server_address_type)
```
Comment (mmlb) [inline installers/osie/main.go:87]: I'm not sure what log_server_address_type means, tcp, udp, ...?

File: installers/osie/mirror.go (hunk 1 above)
Comment (Cbkhare) [reply installers/osie/mirror.go:23]: I followed the existing implementation as in line no. 21.

File: installers/osie/main.go (hunk 2 above)
Comment (Cbkhare) [reply installers/osie/main.go:87]: @mmlb as mentioned in the above description, this PR needs to be reviewed with Tink and Osie Pr.
https://github.com/tinkerbell/tink/pull/267
https://github.com/tinkerbell/osie/pull/104

Comment (Cbkhare) [reply installers/osie/main.go:87]: details of the variable are present here https://github.com/tinkerbell/tink/pull/267/files#diff-95eba583451d93c0a847d4e39967c981

Comment (mmlb) [reply installers/osie/main.go:87]: Gotcha. This info should still be made available in this commit, via the commit message. Ditto for the other PRs.

I'd like for the options to mimic the docker cmdline flags, this helps keeping context of what some variable maps to. Why not `log_opt_syslog-address=log_server_address` instead? Then osie can handle `log_opt*` commands more generically.

File: installers/osie/mirror.go (hunk 1 above)
Comment (mmlb) [reply installers/osie/mirror.go:23]: Can you actually split this off into one item per line. This form of variable declaration isn't really done any where else and should not have been done this way here.

File: installers/osie/main.go (hunk 2 above)
Comment (Cbkhare) [reply installers/osie/main.go:87]: IMO having `syslog` added in env would be incorrect. To show env is part of `log-opt` we can add `opt` in env variables. 
like `LOG_OPT_SERVER_ADDRESS_TYPE`.

File: installers/osie/mirror.go (hunk 1 above)
Comment (Cbkhare) [reply installers/osie/mirror.go]: done

**Comment by Cbkhare** [comment]
https://github.com/tinkerbell/boots/pull/73#discussion_r481340477 @mmlb this comment is addressed

File: installers/osie/main.go [hunk 3]
```diff
@@ -82,6 +82,10 @@ func kernelParams(action, state string, j job.Job, s *ipxe.Script) {
 		s.Args("grpc_cert_url=" + grpcCertURL)
 		s.Args("registry_username=" + registryUsername)
 		s.Args("registry_password=" + registryPassword)
+		s.Args("centralized_logging=" + centralized_logging)
```
Comment (mmlb) [inline installers/osie/main.go:85]: this should not be necessary. The booted machines should be able to figure out if its centralized on its own/may not really care if it configures docker daemon options.

File: installers/osie/mirror.go [hunk 4]
```diff
@@ -14,11 +14,12 @@ const (
...
-	mirrorBaseURL                      = conf.MirrorBaseUrl
-	dockerRegistry                     string
-	grpcAuthority, grpcCertURL         string
-	registryUsername, registryPassword string
+	osieURL                                                              = mustBuildOSIEURL().String()
+	mirrorBaseURL                                                        = conf.MirrorBaseUrl
+	dockerRegistry                                                       string
+	grpcAuthority, grpcCertURL                                           string
+	registryUsername, registryPassword                                   string
+	centralized_logging, log_driver, log_opt_tag, log_opt_server_address string
```
Comment (mmlb) [inline installers/osie/mirror.go:22]: please split to individual lines

File: installers/osie/main.go (hunk 3 above)
Comment (Cbkhare) [reply installers/osie/main.go:85]: `docker logs <service-name>` won't work if there is a log driver defined and eventually this method will throw panic while trying to read the buffer https://github.com/tinkerbell/tink/blob/master/cmd/tink-worker/action.go#L130.

This `centralized_logging` is more of a check if `go captureLogs(ctx, id)`  method should be called or not, https://github.com/tinkerbell/tink/pull/267/files#diff-f776dc055922cd9d519a49a712a9d22fR89

File: installers/osie/mirror.go (hunk 4 above)
Comment (Cbkhare) [reply installers/osie/mirror.go]: done

**Pull request closed without merging**